python_modules/dagster/dagster/components/core/component_hierarchy.py

```diff
@@ -0,0 +1,206 @@
+
+
+def compute_file_hash(path: Path) -> str:
+    return hashlib.sha1(path.read_text().encode("utf-8")).hexdigest()
+
+
+@record
+class ComponentYamlFile:
+    path: Path
+    sha: str
+
+
+@record
+class ComponentDeclaration:
+    path: Path
+
+    @classmethod
+    def build(cls, context: ComponentLoadContext) -> "ComponentDeclaration":
+        return cls(path=context.path)
+
+
+@record
+class YamlComponentDeclaration(ComponentDeclaration):
+    yaml_file: ComponentYamlFile
+
+    @classmethod
+    def build(cls, context: ComponentLoadContext) -> "YamlComponentDeclaration":
+        component_yaml_path = context.path / "component.yaml"
+        return cls(
+            path=context.path,
+            yaml_file=ComponentYamlFile(
+                path=component_yaml_path, sha=compute_file_hash(component_yaml_path)
+            ),
+        )
+
+
+class PythonicComponentDeclaration(ComponentDeclaration): ...
+
+
+class DefsComponentDeclaration(ComponentDeclaration): ...
+
+
+class PythonModuleComponentDeclaration(ComponentDeclaration): ...
+
+
+@record
+class FolderComponentDeclaration(ComponentDeclaration):
+    children: Mapping[Path, ComponentDeclaration]
+    yaml_file: Optional[ComponentYamlFile] = None
+
+    @classmethod
+    def build(cls, context: ComponentLoadContext) -> "FolderComponentDeclaration":
+        found: dict[Path, ComponentDeclaration] = {}
+        for subpath in context.path.iterdir():
+            sub_ctx = context.for_path(subpath)
+            with use_component_load_context(sub_ctx):
+                component = build_component_declaration(sub_ctx)
+                if component:
+                    found[subpath] = component
+
+        component_yaml_path = context.path / "component.yaml"
+        return cls(
+            path=context.path,
+            children=found,
+            yaml_file=None
+            if not component_yaml_path.exists()
+            else ComponentYamlFile(
+                path=component_yaml_path, sha=compute_file_hash(component_yaml_path)
+            ),
+        )
+
+
+@record
+class ComponentHierarchy:
+    root: ComponentDeclaration
+
+
+def build_component_declaration(
+    context: ComponentLoadContext,
+) -> Optional[ComponentDeclaration]:
+    # Check for yaml component
+    component_yaml_path = context.path / "component.yaml"
+    if component_yaml_path.exists():
+        return YamlComponentDeclaration.build(context)
+
+    # Check for pythonic component
+    if (context.path / "component.py").exists():
+        return PythonicComponentDeclaration.build(context)
+
+    # Check for defs component
+    if (context.path / "definitions.py").exists():
+        return DefsComponentDeclaration.build(context)
+
+    # Check for python module
+    if context.path.suffix == ".py":
+        return PythonModuleComponentDeclaration.build(context)
+
+    # Check for folder component
+    if context.path.is_dir():
+        return FolderComponentDeclaration.build(context)
```

Comment:
@alangenfeld is this what you were envisioning?

Reply:
I meant going from `ComponentDeclaration` to `Component` instance down at [1]